From b223ba50fc17af7436b7738f737a5507eee1c6d9 Mon Sep 17 00:00:00 2001
From: Bruce Schmid <duck@freescale.com>
Date: Wed, 18 Jun 2008 15:09:40 -0600
Subject: [PATCH] ENGR00083330-10 Integrate Devtech's USB code patch 10/12

USB IRAM: make sure we don't go off the end of the qtd list.
list_entry(tmp) may return the address of the list head
and not the next QTD, if the current qtd is the last in the list.
This will prevent the use of the list head as a QTD pointer,
which unfortunately causes us to look beyond the end of IRAM.

Signed-off-by: Bruce Schmid <duck@freescale.com>
---
 drivers/usb/host/ehci-q-iram.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/usb/host/ehci-q-iram.c b/drivers/usb/host/ehci-q-iram.c
index afeb1ac..c93ff6b 100644
--- a/drivers/usb/host/ehci-q-iram.c
+++ b/drivers/usb/host/ehci-q-iram.c
@@ -416,7 +416,7 @@ static unsigned qh_completions(struct ehci_hcd *ehci, struct ehci_qh *qh)
 				urb->use_iram = 0;
 				qtd2 =
 				    list_entry(tmp, struct ehci_qtd, qtd_list);
-				if (qtd2) {
+				if (tmp != &qh->qtd_list) {
 					urb2 = qtd2->urb;
 					if (urb2 && urb2->use_iram == 1) {
 						ehci->
-- 
1.5.4.4

